Block-restricted DiracKernel

hugary1995 opened the pull request:
closes #17561

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

hugary1995 committed:
Block-restricted DiracKernel

closes #17561

lindsayad reviewed (commented):
Can you inherit from BlockRestrictable?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/721708/) on d3f6498 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17562/site)

This comment will be updated on new commits.

hugary1995 commented:
Yeah of course -- should've thought about that.

jiangwen84 commented on framework/src/dirackernels/DiracKernel.C:
```diff
@@ -228,7 +231,8 @@ DiracKernel::addPointWithValidId(Point p, unsigned id)
   {
-    const Elem * elem = _dirac_kernel_info.findPoint(p, _mesh);
+    const Elem * elem = blockRestricted() ? _dirac_kernel_info.findPoint(p, _mesh, &blockIDs())
+                                          : _dirac_kernel_info.findPoint(p, _mesh);
```
It might be good to add ` mooseDoOnce(mooseWarning("..."))` to tell users that the point is not found in specified blocks.

hugary1995 replied:
I am not sure if we should do that. Here is a comment from `DiracKernelInfo.C`:
```
  // Note: The PointLocator object returns NULL when the Point is not
  // found within the Mesh.  This is not considered to be an error as
  // far as the DiracKernels are concerned: sometimes the Mesh moves
  // out from the Dirac point entirely and in that case the Point just
  // gets "deactivated".
```
So this predates me -- it has been historically ignored.

lindsayad commented on framework/src/dirackernels/DiracKernel.C:
```diff
@@ -199,7 +201,8 @@ DiracKernel::addPoint(Point p, unsigned id)
   // enabling caching when they add Dirac points.  So all we can do is
   // the PointLocator lookup, and call the other addPoint() method.
-  const Elem * elem = _dirac_kernel_info.findPoint(p, _mesh);
+  const Elem * elem = blockRestricted() ? _dirac_kernel_info.findPoint(p, _mesh, &blockIDs())
+                                        : _dirac_kernel_info.findPoint(p, _mesh);
```
It would be more satisfying to me if we could modify `findPoint` to handle the value of `blockIDs()` returned when we are not `blockRestricted()` (which I believe is `ANY_BLOCK_ID`) then doing these ternary operations everywhere.

lindsayad commented on framework/include/dirackernels/DiracKernelInfo.h:
```diff
@@ -78,7 +79,8 @@ class DiracKernelInfo
    */
-  const Elem * findPoint(Point p, const MooseMesh & mesh);
+  const Elem *
+  findPoint(Point p, const MooseMesh & mesh, const std::set<SubdomainID> * blocks = nullptr);
```
There should be no need for us to modify `p` and consequently not really any desire to copy construct another `Point`, so while we're in here, let's go ahead and change this to `const Point & p`

friedmud commented on framework/src/dirackernels/DiracKernelInfo.C:
```diff
@@ -108,7 +108,7 @@ DiracKernelInfo::updatePointLocator(const MooseMesh & mesh)
 
 const Elem *
-DiracKernelInfo::findPoint(Point p, const MooseMesh & mesh)
+DiracKernelInfo::findPoint(Point p, const MooseMesh & mesh, const std::set<SubdomainID> * blocks)
```
As @lindsayad mentioned - let's not pass blocks by pointer here.  Should be a const reference.

hugary1995 committed:
address some comments

hugary1995 commented:
Thanks @jiangwen84 @lindsayad @friedmud for your review. I have addressed most of your comments. I also changed the signature of `addPoint` and `hasPoint` to pass by const ref.

There are two other things I want to discuss:

1. Regarding @jiangwen84 's comment above, should we just ignore the nullptr returned by the point locator or issue a warning? I see reasons for both behaviors, perhaps add an option for users to choose?
2. Regarding the protected member function `DiracKernel::addPoint(const Elem * elem, Point p, unsigned id = libMesh::invalid_uint)` which is used sort of as a helper function for the two other `addPoint` functions. This function is clearly ignore the block restriction, so it would become users' responsibility to deal with block restrictions if they decide to use it. There are two potential solutions here, I can either add a check in this function, or make this function private. Which one do you suggest?

lindsayad commented:
Are you still waiting for feedback on this @hugary1995 ?

hugary1995 commented:
Yes, I always need your feedback :)

friedmud commented:
For the nullptr return - for DiracKernels I would think that a `mooseWarning` would be appropriate - but I could also see the case for it being a `mooseError` by default - with an input file parameter that allows you to override that.

Yeah - I think that's going to be my vote.  Make it an overridable error.

hugary1995 committed:
Point-not-found behavior

hugary1995 commented:
There are now three behaviors for when a point is not located: error, warning, or ignore. I set the default to be ignore so that this won't be a breaking change.

jiangwen84 reviewed (dismissed):
Seems @hugary1995  addresses all comments.

lindsayad commented on framework/src/dirackernels/DiracKernel.C:
```diff
@@ -38,6 +38,14 @@ DiracKernel::validParams()
       "has been added before. If this option is set to false duplicate points are retained"
       "and contribute to residual and Jacobian.");
 
+  MooseEnum point_not_found_behavior("ERROR WARNING IGNORE", "IGNORE");
+  params.addParam<MooseEnum>(
+      "point_not_found_behavior",
+      point_not_found_behavior,
+      "By default (IGNORE), it is ignored if an added point cannot be located in the "
+      "specified subdomains. If this option is set to ERROR, this situation will result in an "
+      "error. If this option is set to WARNING, then a warning will be issued.");
```
Did you decide to go with `IGNORE` for backwards compatibility? If so then I don't think we've implemented this block-restrictability correctly. I know @friedmud 's vote was for error by default. I think that would be my vote as well. As a potential user I would want to know if my point didn't fall in my subdomain, as opposed to my simulation silently chugging along.

hugary1995 replied:
Marking as resolved because Alex changed his mind -- see below.

lindsayad commented on framework/src/dirackernels/DiracKernel.C:
```diff
@@ -180,7 +190,27 @@ DiracKernel::computeQpOffDiagJacobian(unsigned int /*jvar*/)
+    switch (_point_not_found_behavior)
+    {
+      case PointNotFoundBehavior::ERROR:
+        mooseDoOnce(mooseError(msg.str()));
```
Don't need `mooseDoOnce`

lindsayad commented on framework/src/dirackernels/DiracKernel.C:
```diff
@@ -180,7 +190,27 @@ DiracKernel::computeQpOffDiagJacobian(unsigned int /*jvar*/)
+      case PointNotFoundBehavior::IGNORE:
+        break;
+      default:
+        mooseDoOnce(mooseError("Internal enum error."));
```
Don't need `mooseDoOnce`

lindsayad commented:
>     2\. Regarding the protected member function `DiracKernel::addPoint(const Elem * elem, Point p, unsigned id = libMesh::invalid_uint)` which is used sort of as a helper function for the two other `addPoint` functions. This function is obviously ignoring the block restriction, so it would become users' responsibility to deal with block restrictions if they decide to use it. There are two potential solutions here, I can either add a check in this function, or make this function private. Which one do you suggest?

Did you act on this question? Sorry I've been so focused on other tasks that I've been slow to review your PR. My vote would be to definitely make this `private`

hugary1995 commented:
> > Regarding the protected member function `DiracKernel::addPoint(const Elem * elem, Point p, unsigned id = libMesh::invalid_uint)` which is used sort of as a helper function for the two other `addPoint` functions. This function is obviously ignoring the block restriction, so it would become users' responsibility to deal with block restrictions if they decide to use it. There are two potential solutions here, I can either add a check in this function, or make this function private. Which one do you suggest?
> 
> Did you act on this question? Sorry I've been so focused on other tasks that I've been slow to review your PR. My vote would be to definitely make this `private`

I prefer that as well. [XFEMPressure.C](https://github.com/idaholab/moose/blob/4ee8eed2eb28248797bcdc85e998f33f423ff5fa/modules/xfem/src/dirackernels/XFEMPressure.C#L69) is the only place that uses this function, where the points added are retrieved from a element pair locator. So when we add a point, we skip the geometric search because we already know which element it resides in.

So if we make this function private, we'll have to modify XFEMPressure.C to call the other protected version instead. I don't think that'll break anything, but there is a performance penalty.

hugary1995 commented:
Oh actually `TrackDiracFront` also uses this function. I also don't know how many external and internal apps used it.

Alternatively, I can add a block-restriction check in this DiracKernel::addPoint(const Elem * elem, Point p, unsigned id = libMesh::invalid_uint) function, and also pass in an additional optional boolean (default to true) to turn off the check.

But this `unsigned id = libMesh::invalid_uint` parameter is preventing me from doing that... Any suggestions?

lindsayad commented on framework/src/dirackernels/DiracKernel.C:
```diff
@@ -178,7 +190,27 @@ DiracKernel::computeQpOffDiagJacobian(unsigned int /*jvar*/)
 void
 DiracKernel::addPoint(const Elem * elem, Point p, unsigned /*id*/)
 {
-  if (!elem || (elem->processor_id() != processor_id()))
```
I see now that we silently ignored null elements before so if we do have tests that rely on this, then I'm fine having the default be `IGNORE`

hugary1995 replied:
Yeah, that was relied on by `point_caching_moving_mesh.i`.

hugary1995 commented:
There is yet another issue with making `DiracKernel::addPoint(const Elem * elem, Point p, unsigned id = libMesh::invalid_uint)` private... There will be cases where a point can be located in more than one elements. That is the case with `XFEMPressure.C`.

lindsayad commented:
> Alternatively, I can add a block-restriction check in this `DiracKernel::addPoint(const Elem * elem, Point p, unsigned id = libMesh::invalid_uint)` function, and also pass in an additional optional boolean (default to true) to turn off the check.

Would the check be something like `hasBlocks(elem->subdomain_id())`? I'd be curious if you tried that check whether anything would fail.

Anyway yea let's not make it private. If we can do a sensible check for block restriction that doesn't break any existing behavior then let's do that. If not, let's not bend over backwards for paranoia. At the end of the day historically in MOOSE we have erred on the side of giving more users more flexibility and power, even if that means they shoot themselves in the foot sometimes.

If we cannot put in a sensible check, then just make sure the method has good doxygen

hugary1995 committed:
Address comments, add another block-restriction check

hugary1995 commented:
Okay I think addressed all comments. I would be surprised if anything fails with this extra check.

lindsayad commented:
I will give @friedmud a half day if he wants to review any more. If not, I will merge

hugary1995 commented:
Is @friedmud 's half day over?

lindsayad commented:
Lol
